src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentStack.cs

```diff
@@ -739,8 +739,9 @@ private static IEnumerator<T> GetEnumerator(Node? head)
             Node? current = head;
             while (current != null)
             {
-                yield return current._value;
+                T value = current._value;
                 current = current._next;
+                yield return value;
```

Comment:
@jjonescz, without this change, the tests were consistently failing:
```sh
$ ./build.sh libs+libs.tests
$ ./dotnet.sh build -t:Test src/libraries/System.Collections.Concurrent/tests
...
    Discovering: System.Collections.Concurrent.Tests (method display = ClassAndMethod, method display options = None)
    Discovered:  System.Collections.Concurrent.Tests (found 901 of 909 test cases)
    Starting:    System.Collections.Concurrent.Tests (parallel test collections = on [10 threads], stop on fail = off)
      System.Collections.Concurrent.Tests.ConcurrentStackTests.Enumerator_MoveNext_AfterDisposal(count: 1) [FAIL]
        System.NullReferenceException : Object reference not set to an instance of an object.
        Stack Trace:
          /Users/am11/projects/runtime5/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentStack.cs(743,0): at System.Collections.Concurrent.ConcurrentStack`1.GetEnumerator(Node head)+MoveNext()
          /Users/am11/projects/runtime5/src/libraries/Common/tests/System/Collections/IEnumerable.Generic.Tests.cs(395,0): at System.Collections.Tests.IEnumerable_Generic_Tests`1.Enumerator_MoveNext_AfterDisposal(Int32 count)
             at InvokeStub_IEnumerable_Generic_Tests`1.Enumerator_MoveNext_AfterDisposal(Object, Span`1)
          /Users/am11/projects/runtime5/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.cs(95,0): at System.Reflection.MethodBaseInvoker.InvokeWithOneArg(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
      System.Collections.Concurrent.Tests.ConcurrentStackTests.Enumerator_MoveNext_AfterDisposal(count: 75) [FAIL]
        System.NullReferenceException : Object reference not set to an instance of an object.
        Stack Trace:
          /Users/am11/projects/runtime5/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentStack.cs(743,0): at System.Collections.Concurrent.ConcurrentStack`1.GetEnumerator(Node head)+MoveNext()
          /Users/am11/projects/runtime5/src/libraries/Common/tests/System/Collections/IEnumerable.Generic.Tests.cs(395,0): at System.Collections.Tests.IEnumerable_Generic_Tests`1.Enumerator_MoveNext_AfterDisposal(Int32 count)
             at InvokeStub_IEnumerable_Generic_Tests`1.Enumerator_MoveNext_AfterDisposal(Object, Span`1)
          /Users/am11/projects/runtime5/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.cs(95,0): at System.Reflection.MethodBaseInvoker.InvokeWithOneArg(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
    Finished:    System.Collections.Concurrent.Tests
  === TEST EXECUTION SUMMARY ===
     System.Collections.Concurrent.Tests  Total: 2460, Errors: 0, Failed: 2, Skipped: 0, Time: 74,172s
...
```
is it expected?

Reply:
@jcouv is this due to https://github.com/dotnet/roslyn/pull/75908?

Reply:
Yes, that is the cause.

Here's a [sharplab repro](https://sharplab.io/#v2:C4LgTgrgdgPgsAKAPRIAQCECmALAhgNwEsB7MVAI0wDNTNUBjPKAczoApCpUAiAE2twQANsG4UwuKIwCUIVAAcwnYAGceAFUiYALAGZUvQivnEVmXqlxVgmMpoiZuiFBhwESZKzbKNJrVBxc3AC2uJxi5BJS2LKoAHIASgCizmhJAB7ymPQ2FpR4RKRyispq3PY6+obGpuaoAGK4QmZOCIj4uGQdQg5qALyoUJgA7qgAwgA8ygB8bADa2gBMADSoegC60gDc7Z2omFAQwba4wKSoA929AHQA4pjASYfHEmdgbNuIAAIAjNdjxCgKmIQkw1wA6kobGwDkcTm9rgBZYj4TBxTDpYAfT4IX7/QHA0EQqGYGHPeGkf4QMBgA7AHGIWEvU6UgAiRhMZg+O1xfwBQJBYMhhGhYmqnLq3AZvPxAqJwuhTIpYCRKLRGKx0ml30W4wm6mmiAA3ohUGaFEoOjZUGYmnUvrq4sR+KbzSaEObPahlLYoE1ULTcLxAUIAJ6odSoAD6V0wW1QLgAak0HKhiFRUMBsHQoM6wa6vd6oN4/UJ4nno0NMfGXOjMQpiD6VYgC16m6Xy/w2JHY9JW573YXCzGU3RLqOeUOvVGq8ALoNhEJJ0OAL6ttdtD3mkpWuhO/jR7NB+Mtrdmr76MZsX7aa6/XT66YMEGgnIkKB9s+oQdD/d0IS4CowB/vOhxCEu/bmjQgaMAEkaYKCxzFkWz7gdkwDvp+U4/lOZogUMwwgQMBGdqSCGYEh9LLrhgwjH+1wzhq84AUBf7UbhLHARWxF0Xm7GehukFmlGR4WAMnFseup6eheqA3tcACSTxwq8pCPqg9yPOSqnvFhbpCXJADsGkPMpzJvGwImYEGOICdJ26WqcdC/AAbHJPy3kp2ksmA6maWZypsCBol6WaOFeiB9DUrSyEDKJ/HmoQGZsFFNJ0t6ahgUI0jubKhJCiSbBiBR8jAOGUo8gZwzYIQoIBKlMVzgAhMRi6hV64VTr8PxGQw0V0gxsYJYWDXpQMo3Fgxs7DV6eL8vlxIiqS3BeLYPC2auUkICuQA==).

In the "main" branch, we NRE.
In the "default" branch (ie. before the change to clear locals during disposal), we printed a result without crash. But it looks like that result was incorrect already... 
We were incorrectly allowing enumeration to continue. 

The correct behavior would be for disposal to set the enumerator state to "after" (thereby causing `MoveNext` to return `false` thereafter).
This issue becomes more observable after the change to clear local state (NRE in the present scenario).
Filed [https://github.com/dotnet/roslyn/issues/76078](https://github.com/dotnet/roslyn/issues/76078)

I will discuss with @AlekseyTs @333fred on how best to proceed. I thinking to revert the change temporarily and fix the disposal bug first.

Reply:
@jcouv, thanks for the info. I think we can wait, since there are some other NRE failures in CI which appear to be related.

Reply:
That works, thanks. 
FWIW, here's the [fix in progress](https://github.com/dotnet/roslyn/pull/76090)

Reply:
@am11 The fix was merged into roslyn.

Reply:
@jcouv, thanks for the update. I have updated to latest version. However, this test is still failing without faf8f96b2861b656f905768c054cc87c64d09d62. Were we expecting it to work without that change in the consumer code?

Reply:
@jcouv, thanks for the update. I have updated to latest version. However, `System.Collections.Concurrent.Tests` are still failing without faf8f96b2861b656f905768c054cc87c64d09d62. Were we expecting it to work without that change in the consumer code?

Reply:
It was still the old version. Now it's the one with your fix. It worked! 👍